# Laboratories

AOC Day 7 Part 1

This feels like a mixture of matrices and trees/graphs. Let's bring in our `Matrix` and `Neighbours` implementations from Day 4. I have updated it a bit to fit the data types in the current problem.

In [20]:
class Neighbours:
    def __init__(self, a: str | None, b: str | None, c: str | None, m: str | None, n: str | None, p: str | None, q: str | None, r: str | None) -> None:
        self.a = a
        self.b = b
        self.c = c
        self.m = m
        self.n = n
        self.p = p
        self.q = q
        self.r = r

    # This function returns what's immediately above the current node
    # i.e., the North direction.
    @staticmethod
    def parent(matrix: "Matrix", of: tuple[int, int]) -> str | None:
        return Neighbours.of(matrix=matrix, node=of).b

    @staticmethod
    def of(matrix: "Matrix", node: tuple[int, int]):
        # x = rows, y = cols
        (x, y) = node
        x_min = True if x - 1 < 0 else False
        y_min = True if y - 1 < 0 else False

        x_max = True if x + 1 >= matrix.num_rows() else False
        y_max = True if y + 1 >= matrix.num_cols() else False

        a = None if (x_min or y_min) else matrix[x-1, y-1]
        b = None if x_min else matrix[x -1, y]
        c = None if (x_min or y_max) else matrix[x - 1, y + 1]
        m = None if y_min else matrix[x, y - 1]
        p = None if (y_min or x_max) else matrix[x + 1, y - 1]
        n = None if y_max else matrix[x, y + 1]
        r = None if x_max else matrix[x + 1, y]
        q = None if (y_max or x_max) else matrix[x + 1, y + 1]

        return Neighbours(
            a, b, c, m, n, p, q, r
        )
    
    def to_dict(self):
        return {
            "a": self.a,
            "b": self.b,
            "c": self.c,
            "m": self.m,
            "n": self.n,
            "p": self.p,
            "q": self.q,
            "r": self.r
        }
    
    def __iter__(self):
        yield from [
            self.a,
            self.b,
            self.c,
            self.m,
            self.n,
            self.p,
            self.r,
            self.q
        ]
        
class Matrix:
    def __init__(self, rows: list[str]) -> None:
        self.current_idx = -1
        self.last_idx = len(rows)
        self._vec: list[list] = []
        self._prepare_matrix(rows)
        pass

    def __iter__(self):
        return self
    
    def __next__(self):
        self.current_idx += 1
        if self.current_idx < self.last_idx:
            return self._vec[self.current_idx]

    def _prepare_matrix(self, rows: list[str]) -> None:
        self._vec = [["S" if ch == "S" else  "^" if ch == "^" else "." for ch in line] for line in rows]
        
    def _neighbours(self, node: tuple[int, int]):
        return Neighbours.of(
            matrix=self,
            node=node
        )
    
    def __getitem__(self, key: tuple[int, int]):
        """
        key: (row, column)
        """
        (row, col) = key
        return self._vec[row][col]

    def __setitem__(self, key: tuple[int, int], value: str):
        """
        key: (row, column)
        """
        (row, col) = key
        self._vec[row][col] = value

    def __len__(self):
        return len(self._vec)
    
    def row(self, idx: int) -> list[str]:
        return self._vec[idx]

    def num_rows(self) -> int:
        return len(self._vec)
    
    def num_cols(self) -> int:
        return len(self._vec[0])
    
    def to_string(self) -> str:
        return str(self._vec)
    
    def pretty_print(self) -> None:
        for row in self._vec:
            print(''.join(row))
    
    def dim(self) -> tuple[int, int]:
        return (self.num_rows(), self.num_cols())
    
    def apply_row(self, n, f):
        # Apply function f to an entire row.
        N = self._vec[n]
        for i in range(len(N)):
            f(self, N[i])
    
    def apply(self, f, row_major=True):
        if row_major:
            for i in range(len(self._vec)):
                for j in range(len(self._vec[0])):
                    f(self, self[i, j])
        else:
            for i in range(len(self._vec)):
                for j in range(len(self._vec[0])):
                    f(self, self[j, i])

    def save(self, filename: str):
        """
        Save the matrix to a file, one row per line.
        Each row is joined into a string.
        """
        with open(filename, 'w') as f:
            for row in self._vec:
                f.write(''.join(row) + '\n')


Given below is the input grid.

```
.......S.......
...............
.......^.......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............
```

Given below is the path of the tachyon beam.

```
.......S.......
.......|.......
......|^|......
......|.|......
.....|^|^|.....
.....|.|.|.....
....|^|^|^|....
....|.|.|.|....
...|^|^|||^|...
...|.|.|||.|...
..|^|^|||^|^|..
..|.|.|||.|.|..
.|^|||^||.||^|.
.|.|||.||.||.|.
|^|^|^|^|^|||^|
|.|.|.|.|.|||.|
```

We need to model the behaviour of the tachyon beam first in order to solve this problem.

Let us read the input first.



In [21]:
input = [
    ".......S.......",
    "...............",
    ".......^.......",
    "...............",
    "......^.^......",
    "...............",
    ".....^.^.^.....",
    "...............",
    "....^.^...^....",
    "...............",
    "...^.^...^.^...",
    "...............",
    "..^...^.....^..",
    "...............",
    ".^.^.^.^.^...^.",
    "..............."
]

If we use our `Matrix` implementation to read this into a matrix:

In [22]:
mtx = Matrix(rows=input)

In [23]:
mtx.pretty_print()

.......S.......
...............
.......^.......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............


In [24]:
mtx.dim()

(16, 15)

Looks like the matrix implementation is okay.

We will add some utility functions into the `Matrix` class to traverse through and navigate the grid.

These functions are currently global for testing, but should be moved into the Matrix class after validation. Utility functions for beam navigation in the grid.

In [25]:
def initiate_split(matrix: Matrix):
    """
    Initiate the split from the starting point S.
    
    Note: This mutates the matrix.
    """
    first_row = matrix.row(0)
    start_node_idx = first_row.index("S")
    matrix[1, start_node_idx] = "|"

def quantum_split(matrix: Matrix, row: int):
    """Placeholder for logic to split the beam on the given row in the matrix. n here is the row index."""
    N = matrix.row(row)
    for i in range(len(N)):
        # TODO: We need to check if the beam has actually arrived here.
        if row - 1 >= 0 and matrix[row - 1, i] == "|" and N[i] == "^":
            # split the beam.
            if i - 1 >= 0:
                matrix[row, i - 1] = "|"
            if i + 1 < len(N):
                matrix[row, i + 1] = "|"
        elif row - 1 >= 0 and matrix[row - 1, i] == "|":
            matrix[row, i] = "|"

# Initiate the split.
matrix_cpy = Matrix(input)
initiate_split(matrix_cpy)

for r in range(matrix_cpy.num_rows()):
    quantum_split(matrix=matrix_cpy, row=r)
matrix_cpy.pretty_print()

.......S.......
.......|.......
......|^|......
......|.|......
.....|^|^|.....
.....|.|.|.....
....|^|^|^|....
....|.|.|.|....
...|^|^|||^|...
...|.|.|||.|...
..|^|^|||^|^|..
..|.|.|||.|.|..
.|^|||^||.||^|.
.|.|||.||.||.|.
|^|^|^|^|^|||^|
|.|.|.|.|.|||.|


Looks like we are good!

Let's update the `Matrix` class to our final implementation.

In [26]:
class Matrix:
    def __init__(self, rows: list[str]) -> None:
        self.current_idx = -1
        self.last_idx = len(rows)
        self._vec: list[list] = []
        self._prepare_matrix(rows)
        self._initiate_split()
        self.split_count = 0

    def __iter__(self):
        return self
    
    def __next__(self):
        self.current_idx += 1
        if self.current_idx < self.last_idx:
            return self._vec[self.current_idx]

    def _prepare_matrix(self, rows: list[str]) -> None:
        self._vec = [["S" if ch == "S" else  "^" if ch == "^" else "." for ch in line] for line in rows]

    def __getitem__(self, key: tuple[int, int]):
        """
        key: (row, column)
        """
        (row, col) = key
        return self._vec[row][col]

    def __setitem__(self, key: tuple[int, int], value: str):
        """
        key: (row, column)
        """
        (row, col) = key
        self._vec[row][col] = value

    def __len__(self):
        return len(self._vec)
    
    def _initiate_split(self):
        """
        Initiate the split from the starting point S.
        
        Note: This mutates the matrix.
        """
        first_row = self.row(0)
        start_node_idx = first_row.index("S")
        self[1, start_node_idx] = "|"

    def split_beam(self):
        """
        Splits the beam.
        """
        for row in range(self.num_rows()):
            N = self.row(row)
            for i in range(len(N)):
                if row - 1 >= 0 and self[row - 1, i] == "|" and N[i] == "^":
                    # split the beam.
                    if i - 1 >= 0:
                        self[row, i - 1] = "|"
                    if i + 1 < len(N):
                        self[row, i + 1] = "|"
                    self.split_count += 1
                elif row - 1 >= 0 and self[row - 1, i] == "|":
                    self[row, i] = "|"
    
    def row(self, idx: int) -> list[str]:
        return self._vec[idx]

    def num_rows(self) -> int:
        return len(self._vec)
    
    def num_cols(self) -> int:
        return len(self._vec[0])
    
    def to_string(self) -> str:
        return str(self._vec)
    
    def pretty_print(self) -> None:
        for row in self._vec:
            print(''.join(row))
    
    def dim(self) -> tuple[int, int]:
        return (self.num_rows(), self.num_cols())
    
    def apply_row(self, n, f):
        # Apply function f to an entire row.
        N = self._vec[n]
        for i in range(len(N)):
            f(self, N[i])
    
    def apply(self, f, row_major=True):
        if row_major:
            for i in range(len(self._vec)):
                for j in range(len(self._vec[0])):
                    f(self, self[i, j])
        else:
            for i in range(len(self._vec)):
                for j in range(len(self._vec[0])):
                    f(self, self[j, i])

    def save(self, filename: str):
        """
        Save the matrix to a file, one row per line.
        Each row is joined into a string.
        """
        with open(filename, 'w') as f:
            for row in self._vec:
                f.write(''.join(row) + '\n')

mtx = Matrix(input)
mtx.split_beam()
mtx.pretty_print()

.......S.......
.......|.......
......|^|......
......|.|......
.....|^|^|.....
.....|.|.|.....
....|^|^|^|....
....|.|.|.|....
...|^|^|||^|...
...|.|.|||.|...
..|^|^|||^|^|..
..|.|.|||.|.|..
.|^|||^||.||^|.
.|.|||.||.||.|.
|^|^|^|^|^|||^|
|.|.|.|.|.|||.|


We are good with the `Matrix` class implementation.

In [27]:
mtx.split_count

21

$21$ is in fact, the correct answer!

Let's try to perform the function for our input function.

In [28]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    mtx = Matrix(rows=input)
    mtx.split_beam()
    print(f"Split count = {mtx.split_count}")

    mtx.save("tree.txt")

Split count = 1579


$1579$ is the answer!

---

## Part 2

For part 2, we will have to re-define the splitting behaviour.

In [29]:
input = [
    ".......S.......",
    "...............",
    ".......^.......",
    "...............",
    "......^.^......",
    "...............",
    ".....^.^.^.....",
    "...............",
    "....^.^...^....",
    "...............",
    "...^.^...^.^...",
    "...............",
    "..^...^.....^..",
    "...............",
    ".^.^.^.^.^...^.",
    "..............."
]

matrix = Matrix(rows=input)
matrix.split_beam()
matrix.pretty_print()

.......S.......
.......|.......
......|^|......
......|.|......
.....|^|^|.....
.....|.|.|.....
....|^|^|^|....
....|.|.|.|....
...|^|^|||^|...
...|.|.|||.|...
..|^|^|||^|^|..
..|.|.|||.|.|..
.|^|||^||.||^|.
.|.|||.||.||.|.
|^|^|^|^|^|||^|
|.|.|.|.|.|||.|


### Pascal's Triangle and Plinko

In Stranger Things, there is a game called "Plinko" that they play using a disk and a Galton Board. 

![img](./plinko.gif)

Evidently, the distribution of the disk being eventually collected in a certain column is governed by the Pascal's Triangle.

![img](https://pressbooks.howardcc.edu/app/uploads/sites/61/2020/04/Plinko4-300x252.jpg)

Take $(4,2)$ (using 1-indexing) element (which is $3$) in the Pascal's Triangle. What are the ways that we could arrive at that element starting from the root?

- $1 \rightarrow 1 \rightarrow 1 \rightarrow 3$
- $1 \rightarrow 1 \rightarrow 2 \rightarrow 3$
- $1 \rightarrow 1 \rightarrow 2 \rightarrow 3$ (the second $1$ is the $(2,2)$)

There are $3$ ways to arrive at $(4,2)$, which is equal to $Pascal(4,2)$ (where the function $Pascal(X)$ gives the Pascal's Triangle value of the coordinate $X$) itself!

![img](./pascal.png)

Given above is a small demonstration of how we could calculate the total number of timelines. 

That is, the total number of timelines is equal to the sum of values at the bottom row!

**Note**: Notice how the beams (i.e., `|`) represent nodes in the Pascal's Triangle.

 Because we already have the matrix where we have split the beams, we can write the algorithm above and arrive at the total number of timelines that a tachyon beam is creating.

In [30]:
from functools import reduce

rows, columns = matrix.dim()
rvector = [0] * columns  # Vector of zeros with length equal to number of columns

first_row = matrix.row(0).index("S")
rvector[first_row] = 1

second_row = matrix.row(1).index("|")
rvector[second_row] = 1

for row in range(rows):
    for col in range(columns):
        top_idx = row - 1
        if matrix[row, col] == "^" and top_idx >= 0 and matrix[top_idx, col] == "|":
            # This means we've hit a beam. We've split it.
            if col - 1 >= 0:
                rvector[col - 1] = rvector[col] + rvector[col - 1]
            if col + 1 < columns:
                rvector[col + 1] = rvector[col] + rvector[col + 1]
            rvector[col] = 0

reduce(lambda x, y: y + x, rvector)

40

$40$ is the correct answer for the input. Let's try it for `input.txt`.

In [31]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    matrix = Matrix(rows=input)
    matrix.split_beam()

    rows, columns = matrix.dim()
    rvector = [0] * columns  # Vector of zeros with length equal to number of columns

    first_row = matrix.row(0).index("S")
    rvector[first_row] = 1

    second_row = matrix.row(1).index("|")
    rvector[second_row] = 1

    for row in range(rows):
        for col in range(columns):
            top_idx = row - 1
            if matrix[row, col] == "^" and top_idx >= 0 and matrix[top_idx, col] == "|":
                # This means we've hit a beam. We've split it.
                if col - 1 >= 0:
                    rvector[col - 1] = rvector[col] + rvector[col - 1]
                if col + 1 < columns:
                    rvector[col + 1] = rvector[col] + rvector[col + 1]
                rvector[col] = 0

    print(reduce(lambda x, y: y + x, rvector))


13418215871354


$13418215871354$ is the answer!